Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1

```diff
@@ -473,6 +479,83 @@ function Get-MsiAccessToken {
     }
 }
 
+function Get-SpnAccessTokenUsingCertificate {
+    param(
+        [Parameter(Mandatory=$true)] $endpoint
+    )
+
+    if ($script:certificateAccessToken -and $script:certificateAccessToken.expires_on) {
+        # there exists a token cache
+        $currentTime = [System.DateTime]::UtcNow
+        $tokenExpiryTime = $script:certificateAccessToken.expires_on.UtcDateTime
+        $timeDifference = $tokenExpiryTime - $currentTime
+
+        if ($timeDifference.Minutes -gt 5) {
+            Write-Verbose "Returning access token from cache."
+            return $script:certificateAccessToken
+        }
+    }
+
+    Write-Verbose "Fetching access token using client certificate."
+    
+    # load the ADAL library 
+    Add-Type -Path $PSScriptRoot\Microsoft.IdentityModel.Clients.ActiveDirectory.dll
+
+    $pemFileContent = $endpoint.Auth.Parameters.ServicePrincipalCertificate
+    $pfxFilePath, $pfxFilePassword = ConvertTo-Pfx -pemFileContent $pemFileContent
+    
+    $clientCertificate = Get-PfxCertificate -pfxFilePath $pfxFilePath -pfxFilePassword $pfxFilePassword
+
+    $servicePrincipalId = $endpoint.Auth.Parameters.ServicePrincipalId
+    $tenantId = $endpoint.Auth.Parameters.TenantId
+    $envAuthUrl = $script:defaultEnvironmentAuthUri
+    if($endpoint.Data.environmentAuthorityUrl) {
+        $envAuthUrl = $endpoint.Data.environmentAuthorityUrl
+    }
+
+    $envAuthUrl = Get-EnvironmentAuthUrl -endpoint $endpoint
+    $azureActiveDirectoryResourceId = Get-AzureActiverDirectoryResourceId -endpoint $endpoint
+
+    # add check -- if envAuthUrl ends with adfs then ignore tenantId
+    $authorityUrl = "$envAuthUrl$tenantId"
+    
+    try {
+        $clientAssertionCertificate = New-Object Microsoft.IdentityModel.Clients.ActiveDirectory.ClientAssertionCertificate -ArgumentList $servicePrincipalId, $clientCertificate
+        $authenticationContext = New-Object Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext -ArgumentList $authorityUrl
+        $tokenResult = $authenticationContext.AcquireTokenAsync($azureActiveDirectoryResourceId, $clientAssertionCertificate).ConfigureAwait($false).GetAwaiter().GetResult()
+    }
+    catch {
+        $script:certificateAccessToken = $null
+        throw (Get-VstsLocString -Key "AZ_SPNCertificateAccessTokenFetchFailure" -ArgumentList $servicePrincipalId, $_)
+    }
+
+    if ($tokenResult) {
+        Write-Verbose "Successfully fetched access token using client certificate."
+        $script:certificateAccessToken = @{
+            token_type = $tokenResult.AccessTokenType;
+            access_token = $tokenResult.AccessToken;
+            expires_on = $tokenResult.ExpiresOn;
+        }
+
+        return $script:certificateAccessToken
+    }
+    else {
+        $script:certificateAccessToken = $null
+        throw (Get-VstsLocString -Key "AZ_SPNCertificateAccessTokenFetchFailureTokenNull" -ArgumentList $servicePrincipalId)
+    }
+}
+
+function Get-PfxCertificate {
```

Comment:
There are few functions that are now duplicated between the Initialize Azure and REST helper. Can we refactor?

Reply:
Planning not to refactor for now as there are a few complications.